### PR TITLE
[core] make ProcessFD, [Client,Server]Connection non copyable.

### PR DESCRIPTION
**Pull request by rynewang**
These classes carry system resources (FDs) and are not meant to be copied (and `dup`'d) so to avoid accidental copy we delete the copy constructors.

also removed a dynamic_pointer_cast since it's an upcast so `static` suffices.

### CODE REVIEW

File: src/ray/util/process.cc [hunk 1]
```diff
@@ -299,32 +280,19 @@ ProcessFD &ProcessFD::operator=(ProcessFD &&other) {
   return *this;
 }
 
-intptr_t ProcessFD::CloneFD() const {
```
Comment (rkooo567) [inline src/ray/util/process.cc:302]: it is not used now btw? (I assume yes given deleting copy constructor just succeeds?)

**Review by rkooo567 — APPROVED** [review]

File: src/ray/util/process.cc [hunk 2]
```diff
@@ -299,32 +280,19 @@ ProcessFD &ProcessFD::operator=(ProcessFD &&other) {
...
-    RAY_DCHECK(fd != -1);
-  } else {
-    fd = -1;
+    RAY_CHECK(success) << "error " << errno << " closing process " << pid_ << " FD";
   }
-  return fd;
-}
 
-void ProcessFD::CloseFD() { fd_ = -1; }
+  fd_ = -1;
```
Comment (rkooo567) [inline src/ray/util/process.cc:294]: hmm interseting... this seems actually never used (existing closeFd)

File: src/ray/util/process.cc (hunk 1 above)
Comment (rynewang) [reply src/ray/util/process.cc:302]: it's not used anywhere

File: src/ray/util/process.cc (hunk 2 above)
Comment (rynewang) [reply src/ray/util/process.cc:294]: indeed

**Pull request merged**